### ISSUE TITLE
tutor login via uni id

### ISSUE THREAD

**Issue by dheidemann** [FachschaftMathPhysInfo/pepp#51]
first check whether it is possible to use the auth provider by university

**Comment by dheidemann** [comment]
https://pkg.go.dev/github.com/crewjam/saml

**Comment by fschledorn** [comment]
https://www.urz.uni-heidelberg.de/de/service-katalog/identity-management/shibboleth-identity-provider

Based on the attributes they provide I have written the following list, with attributes we might need. Please feel free to add changes/suggestions to the list.

- displayName: needed if we want to support displayed names to differ from Bürgerlicher Name, and if AD saves those names in this variable (we might need to find this out)
- eduPersonUniqueId: we might need this to use as a UID, we could alternatively use the uid (uni-id)
- givenName: probs needed, in case the display name is not different
- mail: we defo need this 
- persistentID: we could also use this as UID
- sn: we need this
- transientID: there is probs a direct mapping to the SessionID Attribute in the User struct
- uid: we can either use this or the persistendID or the eduPersonUniqueID as UID for the User Database


I currently don’t see a UID in ```server/models/user.go```.  How do you handle identical names so far (ex. Philip Müller) so far? Name conflicts are gonna happen and for shibboleth we defo need a unique ID to map a saml assertion to a specific user. We will also need to make the password and salt attribute optional in the model, or think about how we handle them for SSO.

**Comment by fschledorn** [comment]
Just adding some thoughts below on potential complexities and design questions that came to mind with the Authelia/SAML approach

> I think we should
> 
> 1. think about adding anther table to the database containing third party auth provider-configs. First of all SSO (for authelia) and SAML (for uni)

They work quite differently, right? Authelia seems to be more OIDC or maybe a proxy passing tokens/headers, whereas Shibboleth is the full SAML 2.0 flow with redirects and signed XML assertions. Do you want to handle both? 
How do we want to handle FSler having accounts in both systems, probably with different emails/IDs (persistentID from the uni vs whatever Authelia uses). Do we have them use the FS Account only or is there some other flow?

> > How do you handle identical names so far?
> 
> * `mail` will always be needed as the users unique identifier
> * [pepp/server/models/user.go](https://github.com/FachschaftMathPhysInfo/pepp/blob/bed5b70f878b32453c748566f28b4a1008a425c2/server/models/user.go#L13)   
>            Line 13
>         in
>         [bed5b70](/FachschaftMathPhysInfo/pepp/commit/bed5b70f878b32453c748566f28b4a1008a425c2)         
>              Mail      string    `bun:"type:varchar(255),pk"`
> 
yeah, I saw that. I am not overly familiar with go, so I wasn’t sure if that actually does what I thought. I think the main issue is, that the mail attribute is not static in the universities AD. There may be multiple entries and the one that is given to us via SAML might change without those changes being transparent to us. If my assumption how the university set up their AD is correct, there are no guarantees to the uniques of the mail address over time, which might open us up to a host of attack vectors and headaches. Using a proper UID that is static over the entire lifetime of the account would probably be better in that case. Afaik the persistentID is the standard way to reliably map a user to an account in SAML.
> > We might have to make the password and salt attribute optional in the model
> 
> * they are actually ;)
> * [pepp/server/models/user.go](https://github.com/FachschaftMathPhysInfo/pepp/blob/bed5b70f878b32453c748566f28b4a1008a425c2/server/models/user.go#L19-L20)   
>           Lines 19 to 20
>         in
>         [bed5b70](/FachschaftMathPhysInfo/pepp/commit/bed5b70f878b32453c748566f28b4a1008a425c2)
>              Password  string    `bun:"type:varchar(64)"`         
>              Salt      string    `bun:"type:varchar(22)"`
> 
nice, as I said I’m not overly familiar with Go’s syntax yet, so I apologise for the dumb comment.
> > transientID: there is probs a direct mapping to the SessionID Attribute in the User struct
> 
yeah, I rechecked that. My original comment was wrong. I think I misunderstood the usecause of transientID. It's used to track which authentication request belongs to corresponding saml assertion and afterwards fairly meaningless.

> i think we don't need any other ids than the mail address. 
I guess that depends on how much we want to facilitate things for the users. It might be quite helpful to have the name saved in the unis AD to already be prefilled in the the account instead of everyone having to write it in manually (a bit like muesli or Moodle and so already do). But if we don’t want to support that feature it is not strictly necessary. The only attributes we 100% need are mail, some sort of UID (ex. persistentID) and the transientID to map the requests to the assertions.
>Can be wrong tho as I am not perfectly familiar with SAML yet. But what I think is that its similar to SSO:
> 
> 1. get the user mail provided by the auth provider
> 2. check if user exists
> 3. create user if not
> 4. handle sessions with our own session ids

I think that’s correct from a high level view. To elaborate on the SAML flow the key steps are:

- User Action: User attempts to access our application (SP).
- SP Redirect: SP detects no active session, crafts a SAML AuthnRequest, and redirects the user's browser to the University's IdP (Shibboleth) with the request.
- IdP Authentication: IdP receives the request and authenticates the user (if not already logged in at the IdP).
- IdP Assertion Creation: IdP builds a SAML Assertion containing the agreed-upon stable identifier (persistentID or ePPN) and other requested attributes for the authenticated user.
- IdP Signing: IdP digitally signs the Assertion using its private key to ensure integrity and authenticity.
- IdP Response: IdP sends an HTML form containing the signed SAML Response (with the Assertion inside) back to the user's browser.
- Browser POST: The user's browser automatically POSTs this form to our application's Assertion Consumer Service (ACS) endpoint.
- SP Receives Response: Our application's ACS endpoint receives the SAML Response.
- SP Security Validation: Our application performs critical security checks:
- Verifies the digital signature on the Assertion using the IdP's pre-configured public key.
- Validates Assertion timestamps (NotBefore, NotOnOrAfter) to prevent replay attacks.
- Checks the Audience Restriction to ensure the Assertion is intended for our SP.
- (Other standard checks like Issuer, SubjectConfirmationData recipient may also occur).
- SP Data Extraction: If validation passes, our application securely extracts the user's stable identifier (persistentID/ePPN) and any other attributes from the Assertion.
- SP Internal Logic: This extracted stable identifier is then used as input for our process (Check if user exists -> Create user if not -> Establish application session, ...).

**Comment by dheidemann** [comment]
You are right, the uni ids could cause some headaches... What do you think of the following approach:
We
- add an `id` entity and
- define `PRIMARY_KEY(id, mail)`

for our internal users the `id` can be empty, `mail` is always given. When defining a third party auth, you can map e.g. `persistendID` -> `id`.

> They work quite differently, right?

well, in the end we receive key-value pairs to map on our schema. How we authenticate the platform to the IdP differs indeed between the SSO implementations. What they do to keep their users logged in (sessions, tokens, whatever...) is none of our business, all that matters is we get our attributes parsed

> Do we have them use the FS Account only or is there some other flow

i would say you can do whatever inbox you read more frequently ;)

**Comment by fschledorn** [comment]
> for our internal users the `id` can be empty, `mail` is always given. When defining a third party auth, you can map e.g. `persistendID` -> `id`.

I like that much better. The mail of the “internal” people can be can be more easily kept static. 

> How we authenticate the platform to the IdP differs indeed between the SSO implementations.

I don’t know from the top of my head how much additional work will be required to support both authentication methods but authelia seems to work fine in the test deployment. One thing we will have to keep in mind is that we need to avoid potential collisions of the UID provided by both authelia and shibboleth (if authelia also uses some uid other than mail to uniquely identify users), but that can be easily avoided by some kind of prefix or some other string manipulation would be my guess.

> > Do we have them use the FS Account only or is there some other flow
> 
> i would say you can do whatever inbox you read more frequently ;)

We might have to implement some mechanic to map the accounts to each other. I could see the scenario where a user uses both his Uni account and fs account to create two accounts in pepp. This may lead to some confusion. If this is a non issue for you we can also just ignore that edge case.
